pkg/controller/certificates/podcertificaterequest_controller.go

```diff
@@ -0,0 +1,202 @@
+			},
+		),
+		handler: handler,
+	}
+
+	// Manage the addition/update of certificate requests
+	_, _ = reqInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			req := obj.(*certificatesv1alpha1.PodCertificateRequest)
+			logger.V(4).Info("Adding PodCertificateRequest", "req", req.Name)
+			cc.enqueueRequest(obj)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			oldReq := old.(*certificatesv1alpha1.PodCertificateRequest)
+			logger.V(4).Info("Updating PodCertificateRequest", "old", oldReq.Name)
+			cc.enqueueRequest(new)
+		},
+		DeleteFunc: func(obj interface{}) {
+			req, ok := obj.(*certificatesv1alpha1.PodCertificateRequest)
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					logger.V(2).Info("Couldn't get object from tombstone", "object", obj)
+					return
+				}
+				req, ok = tombstone.Obj.(*certificatesv1alpha1.PodCertificateRequest)
+				if !ok {
+					logger.V(2).Info("Tombstone contained object that is not a PodCertificateRequest", "object", obj)
+					return
+				}
+			}
+			logger.V(4).Info("Deleting PodCertificateRequest", "req", req.Name)
+			cc.enqueueRequest(obj)
+		},
+	})
+	cc.reqLister = reqInformer.Lister()
+	cc.reqsSynced = reqInformer.Informer().HasSynced
+	return cc
+}
+
+// Run the main goroutine responsible for watching and syncing jobs.
+func (cc *PodCertificateRequestController) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer cc.queue.ShutDown()
+
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting certificate controller", "name", cc.name)
+	defer logger.Info("Shutting down certificate controller", "name", cc.name)
+
+	if !cache.WaitForNamedCacheSync(fmt.Sprintf("certificate-%s", cc.name), ctx.Done(), cc.reqsSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, cc.worker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+// worker runs a thread that dequeues CSRs, handles them, and marks them done.
+func (cc *PodCertificateRequestController) worker(ctx context.Context) {
+	for cc.processNextWorkItem(ctx) {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (cc *PodCertificateRequestController) processNextWorkItem(ctx context.Context) bool {
+	cKey, quit := cc.queue.Get()
+	if quit {
+		return false
+	}
+	defer cc.queue.Done(cKey)
+
+	if err := cc.syncFunc(ctx, cKey); err != nil {
+		cc.queue.AddRateLimited(cKey)
+		var ign ignorableError
+		if ignorable := errors.As(err, &ign); !ignorable {
+			utilruntime.HandleError(fmt.Errorf("sync %v failed: %w", cKey, err))
+		} else {
+			klog.FromContext(ctx).V(4).Info("Sync certificate request failed", "csr", cKey, "err", err)
+		}
+		return true
+	}
+
+	cc.queue.Forget(cKey)
+	return true
+
+}
+
+func (cc *PodCertificateRequestController) enqueueRequest(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
```

Comment:
shouldn't we enqueue only the objects for the signer we care about?

Reply:
Ditto, this was dropped from this PR.